src/components/tl/ucp/allgather/allgather_knomial.c

```diff
@@ -234,6 +262,183 @@ ucc_status_t ucc_tl_ucp_allgather_knomial_start(ucc_coll_task_t *coll_task)
+    void                  *sbuf;
+    ptrdiff_t              peer_seg_offset, local_seg_offset;
+    ucc_rank_t             peer, peer_dist;
+    ucc_kn_radix_t         loop_step;
+    size_t                 peer_seg_count, local_seg_count;
+    ucc_status_t           status;
+    size_t                 extra_count;
+
+    ucc_tl_ucp_context_t *ctx = UCC_TL_UCP_TEAM_CTX(team);
+    ucp_mem_map_params_t mmap_params;
+    ucp_mem_h            mh;
+    int size_of_list = 1;
+    int count_mh = 0; 
+    ucp_mem_h *mh_list = (ucp_mem_h *)malloc(size_of_list * sizeof(ucp_mem_h));
+
+    mmap_params.field_mask  = UCP_MEM_MAP_PARAM_FIELD_ADDRESS |
+                            UCP_MEM_MAP_PARAM_FIELD_LENGTH  |
+                            UCP_MEM_MAP_PARAM_FIELD_MEMORY_TYPE;
+    mmap_params.memory_type = ucc_memtype_to_ucs[mem_type];
+
+    if (KN_NODE_EXTRA == node_type) {
+        if (p->type != KN_PATTERN_ALLGATHERX) {
+
+            mmap_params.address     = task->allgather_kn.sbuf;
+            mmap_params.length      = local * dt_size;
+            status = ucp_mem_map(ctx->worker.ucp_context, &mmap_params, &mh);
+            if (UCC_OK != status) {
+                task->super.status = status;
+                return;
+            }
+            if (count_mh == size_of_list){
+                size_of_list *= 2;
+                mh_list = (ucp_mem_h *)realloc(mh_list, size_of_list * sizeof(ucp_mem_h));
+            }
+            mh_list[count_mh++] = mh;
+        }
+        
+        mmap_params.address     = rbuf;
+        mmap_params.length      = data_size;
+        status = ucp_mem_map(ctx->worker.ucp_context, &mmap_params, &mh);
+        if (UCC_OK != status) {
+            task->super.status = status;
+            return;
+        }
+        if (count_mh == size_of_list){
+            size_of_list *= 2;
+            mh_list = (ucp_mem_h *)realloc(mh_list, size_of_list * sizeof(ucp_mem_h));
+        }
+        mh_list[count_mh++] = mh;
+    }
+    if ((p->type != KN_PATTERN_ALLGATHERX) && (node_type == KN_NODE_PROXY)) {
+        peer = ucc_knomial_pattern_get_extra(p, rank);
+        extra_count = GET_LOCAL_COUNT(args, size, peer);
+        peer = ucc_ep_map_eval(task->subset.map, peer);
+        mmap_params.address     = PTR_OFFSET(task->allgather_kn.sbuf,
+                                        local * dt_size);
+        mmap_params.length      = extra_count * dt_size;
+        status = ucp_mem_map(ctx->worker.ucp_context, &mmap_params, &mh);
+        if (UCC_OK != status) {
+            task->super.status = status;
+            return;
+        }
+        if (count_mh == size_of_list){
+            size_of_list *= 2;
+            mh_list = (ucp_mem_h *)realloc(mh_list, size_of_list * sizeof(ucp_mem_h));
+        }
+        mh_list[count_mh++] = mh;
+    }
+
+    if ((KN_NODE_EXTRA == node_type) || (KN_NODE_PROXY == node_type)) {
+        if (KN_NODE_EXTRA == node_type) {
+            goto out;
```

Comment:
there shouldn't be any use of OUT in register_memory, its needed only in progress function, not here which is blocking

Reply:
but before there was this :

UCC_KN_PHASE_EXTRA:
    if ((KN_NODE_EXTRA == node_type) || (KN_NODE_PROXY == node_type)) {
        if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
            SAVE_STATE(UCC_KN_PHASE_EXTRA);
            return;
        }
        if (KN_NODE_EXTRA == node_type) {
            goto out;
        }
    }
    
    I removed everything and you told me that the second part was actually needed

Reply:
here : https://github.com/openucx/ucc/pull/1010/commits/e14d2b16d16145a4467126130488e7120d3bfd60#r1727153646